### PR TITLE
New -node-containerd-registry-mirrors flag

### PR DESCRIPTION
**Pull request by kron4eg**
**What this PR does / why we need it**:
This is a companion/dependency PR for https://github.com/kubermatic/kubeone/pull/1674

**Special notes for your reviewer**:

**Optional Release Note**:

```release-note
New -node-containerd-registry-mirrors flag
```

### CODE REVIEW

File: cmd/machine-controller/main.go [hunk 1]
```diff
@@ -253,10 +295,23 @@ func main() {
 	var registryMirrors []string
 	for _, mirror := range strings.Split(nodeRegistryMirrors, ",") {
 		if trimmedMirror := strings.TrimSpace(mirror); trimmedMirror != "" {
+			if !strings.HasPrefix(mirror, "http") {
+				trimmedMirror = "https://" + mirror
```
[the remark this answers is not in the document]
Comment (kron4eg) [reply cmd/machine-controller/main.go:299]: https://docs.docker.com/registry/insecure/

> With insecure registries enabled, Docker goes through the following steps:
> * First, try using HTTPS.
>   * If HTTPS is available but the certificate is invalid, ignore the error about the certificate.
>   * If HTTPS is not available, fall back to HTTP.

And so, by default (in insecure is not specified and no explicit http://) the registry is considered secure, hence is -- https://

File: cmd/machine-controller/main.go [hunk 2]
```diff
@@ -83,12 +85,51 @@ var (
...
+	sort.Strings(registryNames)
+
+	for _, registryName := range registryNames {
+		for _, mirror := range fl[registryName] {
+			result = append(result, fmt.Sprintf("%s=%s", registryName, mirror))
+		}
+	}
+
+	return fmt.Sprintf("%v", result)
+}
```
[the remark this answers is not in the document]
Comment (kron4eg) [reply cmd/machine-controller/main.go]: :+1: :heavy_check_mark:

File: cmd/machine-controller/custom_flags.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	var (
+		registryNames []string
+		result        []string
+	)
+
+	for registryName := range fl {
+		registryNames = append(registryNames, registryName)
+	}
+
+	sort.Strings(registryNames)
```
Comment (xmudrii) [inline cmd/machine-controller/custom_flags.go:51]: Why is sorting required?

Comment (kron4eg) [reply cmd/machine-controller/custom_flags.go:51]: yes, to have stable output instead of randomized

File: cmd/machine-controller/main.go [hunk 4]
```diff
@@ -253,10 +258,23 @@ func main() {
...
+				klog.Fatalf("incorrect mirror provided: %v", err)
+			}
+
 			registryMirrors = append(registryMirrors, trimmedMirror)
 		}
 	}
 
+	if len(registryMirrors) > 0 {
+		nodeContainerdRegistryMirrors["docker.io"] = registryMirrors
+	}
```
Comment (xmudrii) [inline cmd/machine-controller/main.go:276]: Should we have some validation to fail starting MC if both `registryMirrors` and `nodeContainerdRegistryMirrors["docker.io"]` are provided by the user?

Comment (kron4eg) [reply cmd/machine-controller/main.go:276]: No, if docker.io is empty it's OK for dockerd

File: cmd/machine-controller/main.go [hunk 5]
```diff
@@ -163,6 +167,7 @@ func main() {
 	flag.StringVar(&nodePauseImage, "node-pause-image", "", "Image for the pause container including tag. If not set, the kubelet default will be used: https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/")
 	flag.StringVar(&nodeKubeletRepository, "node-kubelet-repository", "quay.io/kubermatic/kubelet", "Repository for the kubelet container. Only has effect on Flatcar Linux.")
 	flag.StringVar(&nodeContainerRuntime, "node-container-runtime", "docker", "container-runtime to deploy")
+	flag.Var(&nodeContainerdRegistryMirrors, "node-containerd-registry-mirrors", "Configure registry mirrors endpoints. Can be used multiple times to specify multiple mirrors")
```
Comment (embik) [inline cmd/machine-controller/main.go:170]: It seems I can't add a comment to the line for `node-registry-mirrors` above this one, but should we update the description for `node-registry-mirrors` to either call it deprecated or add that it only is effective for `docker.io` images?

File: cmd/machine-controller/main.go (hunk 4 above)
Comment (xmudrii) [reply cmd/machine-controller/main.go:276]: I'm not sure I get it, but isn't one flag going to override another one? For example, if you have both `--node-registry-mirrors` and `--node-containerd-registry-mirrors`, isn't only one going to have effect?

File: cmd/machine-controller/main.go (hunk 5 above)
Comment (kron4eg) [reply cmd/machine-controller/main.go:170]: it's not depricated :)

File: cmd/machine-controller/main.go (hunk 4 above)
Comment (kron4eg) [reply cmd/machine-controller/main.go:276]: Essentially it will, but I don't see it as a problem.

File: cmd/machine-controller/main.go (hunk 5 above)
Comment (embik) [reply cmd/machine-controller/main.go:170]: Gotcha, wasn't sure what status we considered for it.

File: cmd/machine-controller/main.go (hunk 4 above)
Comment (kron4eg) [reply cmd/machine-controller/main.go:276]: It's kind of backward compatibility, it's possible to configure docker and containerd at the same time (of course only 1 will be used, according to -node-container-runtime).

Comment (xmudrii) [reply cmd/machine-controller/main.go:276]: Okay, that makes sense, thanks for the explanation!

**Review by xmudrii — APPROVED** [review]

**Pull request merged**